src/main/java/hudson/plugins/tfs/TeamBuildEndpoint.java

```diff
@@ -0,0 +1,223 @@
+package hudson.plugins.tfs;
+
+import hudson.Extension;
+import hudson.model.AbstractProject;
+import hudson.model.Item;
+import hudson.model.UnprotectedRootAction;
+import hudson.plugins.tfs.model.AbstractCommand;
+import hudson.plugins.tfs.model.BuildCommand;
+import hudson.plugins.tfs.model.PingCommand;
+import hudson.plugins.tfs.util.MediaType;
+import jenkins.model.Jenkins;
+import jenkins.util.TimeDuration;
+import net.sf.json.JSONObject;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringEscapeUtils;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.HttpResponses;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.Collections;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+import static javax.servlet.http.HttpServletResponse.SC_METHOD_NOT_ALLOWED;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+
+/**
+ * The endpoint that TFS/Team Services will PUT to when it wants to schedule a build in Jenkins.
+ */
+@Extension
+public class TeamBuildEndpoint implements UnprotectedRootAction {
+
+    private static final Logger LOGGER = Logger.getLogger(TeamBuildEndpoint.class.getName());
+    private static final Map<String, AbstractCommand.Factory> COMMAND_FACTORIES_BY_NAME;
+    public static final String URL_NAME = "team-build";
+    static final String URL_PREFIX = "/" + URL_NAME + "/";
+
+    static {
+        final Map<String, AbstractCommand.Factory> map = new TreeMap<String, AbstractCommand.Factory>(String.CASE_INSENSITIVE_ORDER);
+        map.put("ping", new PingCommand.Factory());
+        map.put("build", new BuildCommand.Factory());
+        COMMAND_FACTORIES_BY_NAME = Collections.unmodifiableMap(map);
+    }
+
+    private String commandName;
+    private String jobName;
+
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @Override
+    public String getUrlName() {
+        return URL_NAME;
+    }
+
+    public String getCommandName() {
+        return commandName;
+    }
+
+    public String getJobName() {
+        return jobName;
+    }
+
+    boolean decodeCommandAndJobNames(final String pathInfo) {
+        if (pathInfo.startsWith(URL_PREFIX)) {
+            final String restOfPath = pathInfo.substring(URL_PREFIX.length());
+            final int firstSlash = restOfPath.indexOf('/');
+            if (firstSlash != -1) {

```

Comment:
> What if firstSlash == 0?

Interesting scenario.  `firstSlash` will be `0` happens when the input looks like `/team-build//ping/a` and therefore the `commandName` is decoded as the empty string, while the `jobName` is decoded as `ping/a`.  

The caller (`dispatch()`) responds with HTTP 400 (Bad request) saying "Command not implemented" as a result.